src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -218,58 +212,46 @@ public void Add(string key, object value)
         {
             if (key == null)
             {
-                throw new ArgumentNullException(nameof(key));
+                throw new ArgumentNullException("key");
             }
 
-            Upgrade();
+            EnsureCapacity(_count + 1);
 
-            var list = (ListStorage)_storage;
-            for (var i = 0; i < list.Count; i++)
+            var index = FindArray(key);
+            if (index >= 0)
             {
-                if (string.Equals(list[i].Key, key, StringComparison.OrdinalIgnoreCase))
-                {
-                    var message = Resources.FormatRouteValueDictionary_DuplicateKey(key, nameof(RouteValueDictionary));
-                    throw new ArgumentException(message, nameof(key));
-                }
+                var message = Resources.FormatRouteValueDictionary_DuplicateKey(key, nameof(RouteValueDictionary));
+                throw new ArgumentException(message, "key");
             }
 
-            list.Add(new KeyValuePair<string, object>(key, value));
+            _arrayStorage[_count] = new KeyValuePair<string, object>(key, value);
```

Comment:
what happens here if dictionary is in _propertyStorage mode

Reply:
EnsureCapacity will make it writable before you get here